### PR TITLE
Add Concurrency for CI

### PR DESCRIPTION
**Pull request by slaymance**
#### Description of changes
Adds [concurrency](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency) condition for our CI workflows. This change will define a group based on the pull request ID and whenever a new CI job is triggered for that group (e.g. a new commit on the PR) any in-progress workflows in that group will be canceled.

This should help alleviate some pressure on the number of concurrent GitHub Action runners we utilize.

#### Issue #, if available
N/A

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated - N/A
- [x] Relevant documentation is changed or added (and PR referenced) - N/A

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -3,6 +3,10 @@
 
 name: Tests
 
+concurrency:
+  group: ${{ github.event.pull_request.id }}
```
Comment (wlee221) [inline .github/workflows/tests.yml:7]: How does this work for `on: push`?

Comment (slaymance) [reply .github/workflows/tests.yml]: Good callout, I followed the advice for [adding a fallback value](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-fallback-value).

Comment (wlee221) [reply .github/workflows/tests.yml]: Oh man, I didn't know `||` could be used as fallback operator just like in JS! This is nice.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -3,6 +3,10 @@
 
 name: Tests
 
+concurrency:
+  group: ${{ github.event.pull_request.id || github.ref }}
```
Comment (wlee221) [inline .github/workflows/tests.yml:7]: Sorry for nits after another :D Let's specify this group name, so it doesn't coincide with other concurrency groups we might add later.

```suggestion
  group: e2e-${{ github.event.pull_request.id || github.ref }}
```

**Pull request merged**